### PR TITLE
[CDAP-17151, CDAP-16997] Add submitTime in the PreviewStatus for UI and reconcile the preview run status if pod dies because of OOM.

### PR DESCRIPTION
**Pull request by sagarkapare**
This PR contains following changes in separate commits:
Commit 1: Added submitTime to PreviewStatus along with existing startingTime so UI can display clock correctly on page refreshes. https://issues.cask.co/browse/CDAP-17151

Commit 2: When preview runner pod starts executing  a preview, it stores the program id on a disk attached to the pod. At the end of preview run (successfully or otherwise) file containing program id is deleted. As a part of bootstrap process, preview runner pod checks if any program id file exists from the previous run, if it does it means pod was terminated abruptly without having an opportunity to send the preview status to preview manager. Send RUN_FAILED status for such preview. https://issues.cask.co/browse/CDAP-16997

### CODE REVIEW

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java [hunk 1]
```diff
@@ -51,18 +51,20 @@ public boolean isEndState() {
 
   private final Status status;
   private final BasicThrowable throwable;
+  private final Long submitTime;
   private final Long startTime;
   private final Long endTime;
   private final Integer positionInWaitingQueue;
 
-  public PreviewStatus(Status status, @Nullable BasicThrowable throwable, @Nullable Long startTime,
+  public PreviewStatus(Status status, Long submitTime, @Nullable BasicThrowable throwable, @Nullable Long startTime,
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java]: Why use `Long` instead of the primitive `long`?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/preview/PreviewHttpHandler.java [hunk 2]
```diff
@@ -209,14 +209,19 @@ public void getPreviewLogsNext(HttpRequest request, HttpResponder responder,
...
                             fromOffsetStr, escape, filterStr, info.getRunRecord(), format, suppress));
   }
 
-  private void sendLogs(String namespaceId, String previewId, Consumer<LogReaderInfo> logsResponder) throws Exception {
+  private void sendLogs(HttpResponder responder, String namespaceId, String previewId,
+                        Consumer<LogReaderInfo> logsResponder) throws Exception {
     ApplicationId applicationId = new ApplicationId(namespaceId, previewId);
     ProgramRunId programRunId = previewManager.getRunId(applicationId);
+    if (programRunId == null) {
+      responder.sendStatus(HttpResponseStatus.OK);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/preview/PreviewHttpHandler.java:222]: Why it is OK instead of NOT_FOUND? The UI can't handle it correctly?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java [hunk 3]
```diff
@@ -117,7 +125,8 @@
                        ProgramNotificationSubscriberService programNotificationSubscriberService,
                        LevelDBTableService levelDBTableService,
                        StructuredTableAdmin structuredTableAdmin,
-                       StructuredTableRegistry structuredTableRegistry) {
+                       StructuredTableRegistry structuredTableRegistry, PreviewRunWrapper previewRunWrapper,
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java]: Having a class name with `Wrapper` at the top level is awkward. Typically wrapper class should be private/internal. Why we need this interface? In standalone we can write out the same file and delete it, isn't it?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java (hunk 1 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java]: done.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java (hunk 3 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java]: Thanks for the explanation! Updated code to remove the `PreviewRunWrapper` interface since it was not used in Standalone. 
Added `DistributedPreviewRunner` which extends from the `DefaultPreviewRunner` and does the job of file handling.
Please check again.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/preview/PreviewHttpHandler.java (hunk 2 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/preview/PreviewHttpHandler.java:222]: Returning `OK` is similar behavior from 6.1.3 https://github.com/cdapio/cdap/blob/v6.1.3/cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/preview/PreviewHttpHandler.java#L227 as well. I updated to return `NOT_FOUND` in between, but seems like its temporary state till the run id is propagated back from preview runner. UI shows NOT_FOUND during that time. Reverting it to make is same as 6.1.3

**Comment by sagarkapare** [comment]
Build is green - https://builds.cask.co/browse/CDAP-RUT1838-5

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/DistributedPreviewRunner.java [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        ProgramId programId = GSON.fromJson(reader, ProgramId.class);
+        long submitTimeMillis = RunIds.getTime(programId.getApplication(), TimeUnit.MILLISECONDS);
+        PreviewStatus status = new PreviewStatus(
+          PreviewStatus.Status.RUN_FAILED, submitTimeMillis,
+          new BasicThrowable(new Exception("Preview runner container killed possibly because of out of memory. " +
+                                             "Please try running preview again.")),
+          null, null);
+        setStatus(programId, status);
+      }
+      destroy();
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/DistributedPreviewRunner.java]: Will calling destroy() transit this guava `Service` to COMPLETED state?? Do you want this service stop?

Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/DistributedPreviewRunner.java]: No `destroy()` just deletes the left over file from previous preview run on line 116 here - https://github.com/cdapio/cdap/pull/12589/files/a9a6ec661469ccafa8e3aa3ee77b462040882b2b#diff-fe8b7d5449ae6f60f24802aa2104bd47R116

It wont cause any service state transition. 
Should I rename `init` to `onPreviewStart` and `destroy` to `onPreviewEnd` for clarity?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java [hunk 5]
```diff
@@ -73,6 +75,10 @@ public Status getStatus() {
     return status;
   }
 
+  public Long getSubmitTime() {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java]: `long`?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java [hunk 6]
```diff
@@ -116,6 +122,7 @@ public boolean equals(Object o) {
 
     PreviewStatus status1 = (PreviewStatus) o;
     return status == status1.status &&
+      Objects.equals(submitTime, status1.submitTime) &&
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java]: For primitive type, use `==` to compare

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java [hunk 7]
```diff
@@ -132,14 +144,19 @@
     this.levelDBTableService = levelDBTableService;
     this.structuredTableAdmin = structuredTableAdmin;
     this.structuredTableRegistry = structuredTableRegistry;
+    this.previewIdDirPath = Paths.get(cConf.get(Constants.CFG_LOCAL_DATA_DIR), "previewid").toAbsolutePath();
+    try {
+      Files.createDirectories(previewIdDirPath);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java]: Should be done in the startup method and let the exception bubble out. Generally not good to throw exception from constructor for classes that are created via Guice injection.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java [hunk 8]
```diff
@@ -264,7 +285,7 @@ private void trackPreviewTimeout(PreviewRequest previewRequest, AtomicBoolean ti
     timeoutThread.start();
   }
 
-  private void setStatus(ProgramId programId, PreviewStatus previewStatus) {
+  protected void setStatus(ProgramId programId, PreviewStatus previewStatus) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java]: Is this change still necessary?

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java [hunk 9]
```diff
@@ -299,6 +320,27 @@ protected void startUp() throws Exception {
       metricsCollectionService.start(),
       programNotificationSubscriberService.start()
     ).get();
+
+    // Reconcile status for abruptly terminated preview runs
+    try (Stream<Path> paths = Files.walk(Paths.get(previewIdDirPath.toString()))) {
+      paths.forEach(path -> {
+        if (Files.isDirectory(path)) {
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java]: Use `paths.filter(Files::isRegularFile).forEach`

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java [hunk 10]
```diff
@@ -299,6 +320,27 @@ protected void startUp() throws Exception {
...
+          ProgramId programId = GSON.fromJson(reader, ProgramId.class);
+          long submitTimeMillis = RunIds.getTime(programId.getApplication(), TimeUnit.MILLISECONDS);
+          PreviewStatus status = new PreviewStatus(
+            PreviewStatus.Status.RUN_FAILED, submitTimeMillis,
+            new BasicThrowable(new Exception("Preview runner container killed possibly because of out of memory. " +
+                                               "Please try running preview again.")),
+            null, null);
+          previewTerminated(programId, status);
+        } catch (IOException e) {
+          LOG.warn("Error reading file {}. Ignoring", path.toString(), e);
```
Comment (chtyim) [inline cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java]: No need to call `.toString`

Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java]: ack!

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java (hunk 9 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java]: done.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java (hunk 8 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java]: no its not required now. reverted.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java (hunk 7 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRunner.java]: make sense! moved to startup.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java (hunk 6 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java]: done.

File: cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java (hunk 5 above)
Comment (sagarkapare) [reply cdap-app-fabric/src/main/java/io/cdap/cdap/app/preview/PreviewStatus.java]: right. fixed.

**Review by chtyim — APPROVED** [review]

**Comment by sagarkapare** [comment]
Build is green here - https://builds.cask.co/browse/CDAP-RUT1838-11. Squashing and merging.
